### PR TITLE
use ((->) Unit) instead of Lazy

### LINKED ISSUE #83: Provide non memoized Trampoline?
Currently `Trampoline` is defined in terms of `Lazy` which is memoized. Because of this `Trampoline` is also memoized. But there could also exist non memoized definition for `Trampoline` using just `((->) Unit)` (Currently `->` is not `MonadRec`, but will become once https://github.com/purescript/purescript-tailrec/pull/18 is merged). This implementation might be more efficient as it has no extra Lazy wrapper and memoization. 

We have two options:

1. use `->` instead of `Lazy` in `Trampoline`
2. define new Trampoline in terms of `->`

### PR DESCRIPTION
**Pull request by safareli**
fix #83

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -7,7 +7,9 @@
   },
   "devDependencies": {
     "pulp": "^10.0.4",
-    "purescript-psa": "^0.5.0-rc.1",
+    "pulp": "^11.0.x",
+    "purescript": "0.11.x",
+    "purescript-psa": "^0.5.0",
```
Comment (safareli) [inline package.json]: are this changes fine?

File: src/Control/Monad/Trampoline.purs [hunk 2]
```diff
@@ -31,14 +29,10 @@ done = pure
 suspend :: forall a. Trampoline a -> Trampoline a
 suspend = suspendF
```
Comment (safareli) [inline src/Control/Monad/Trampoline.purs]: I don't quite understand what's the point of `suspendF` and `suspend`. does it make sense for this implementation?

Comment (paf31) [inline src/Control/Monad/Trampoline.purs]: I think we can remove it since this is breaking anyway.

**Comment by safareli** [comment]
btw `/benchmark` is outdated and i was not able to run it using PS@0.11

File: package.json (hunk 1 above)
Comment (safareli) [inline package.json]: by adding ps here we don't need to fetch it by hand in build

File: .travis.yml [hunk 3]
```diff
@@ -2,13 +2,7 @@ language: node_js
...
-env:
-  - PATH=$HOME/purescript:$PATH
 install:
-  - TAG=$(wget -q -O - https://github.com/purescript/purescript/releases/latest --server-response --max-redirect 0 2>&1 | sed -n -e 's/.*Location:.*tag\///p')
-  - wget -O $HOME/purescript.tar.gz https://github.com/purescript/purescript/releases/download/$TAG/linux64.tar.gz
-  - tar -xvf $HOME/purescript.tar.gz -C $HOME/
-  - chmod a+x $HOME/purescript
   - npm install -g bower
   - npm install
 script:
```
Comment (safareli) [inline .travis.yml]: What's the point of two bower Installs here?

File: package.json (hunk 1 above)
Comment (garyb) [inline package.json]: For the core libraries we prefer to fetch by hand - it means that we can update libraries before the `npm` release is done.

File: .travis.yml (hunk 3 above)
Comment (garyb) [inline .travis.yml]: The second `bower install` will fetch `devDependencies`, splitting the build/test/install stuff this way ensures we're not accidentally relying on dev dependencies in the library.

**Comment by ethul** [comment]
This may already be in the works or may have already been decided; however, I am wondering if it would make sense to add a performance comparison to test the note in #83 that the implementation in this PR might be more efficient.

**Comment by safareli** [comment]
@ethul I would need to update `/benchmark` as it's outdated. will try

**Comment by ethul** [comment]
Thanks!

**Comment by natefaubion** [comment]
Is the Lazy/thunk even necessary? Why isn't the spine of `Free Identity` enough to act as a trampoline?

**Comment by natefaubion** [comment]
Example: http://try.purescript.org/?backend=core&gist=f89133003a49095821f592f0dc2b8c16

**Comment by natefaubion** [comment]
FWIW, I'm not sure that `Free Lazy` memoizes in a very interesting way. If you defer under a monadic bind, it won't share anything because it creates a new `Lazy` every time you apply the bind.

**Comment by paf31** [comment]
It's useful to form cyclic structures, and I think some could be generated purely using `Lazy` which would then share some results. Things like finding comonadic fixed points jump to mind. That might be useful, but whether it's useful enough to add the performance overhead for the simple case, I don't know.

**Comment by safareli** [comment]
What if we just remove it as there are already 3 ways to get trampoline? (`Lazy`, `(->) Unit`, `Identity`)

**Comment by paf31** [comment]
Uh, sorry, I was thinking of `Cofree`, not `Trampoline`.

**Comment by safareli** [comment]
> What if we just remove it as there are already 3 ways to get trampoline? (Lazy, (->) Unit, Identity)

?

**Pull request merged**